extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -44,47 +48,48 @@ public class HttpAuthenticator {
     private final HttpAuthenticationMechanism[] mechanisms;
     private final SecurityEventHelper<AuthenticationSuccessEvent, AuthenticationFailureEvent> securityEventHelper;
 
-    public HttpAuthenticator(IdentityProviderManager identityProviderManager,
-            Event<AuthenticationFailureEvent> authFailureEvent,
-            Event<AuthenticationSuccessEvent> authSuccessEvent,
-            BeanManager beanManager,
-            Instance<HttpAuthenticationMechanism> httpAuthenticationMechanism,
+    HttpAuthenticator(IdentityProviderManager identityProviderManager,
             Instance<IdentityProvider<?>> providers,
+            InjectableInstance<BasicAuthenticationMechanism> basicAuthMechanismInstance,
+            Event<AuthenticationFailureEvent> authFailureEvent, Event<AuthenticationSuccessEvent> authSuccessEvent,
+            BeanManager beanManager, @All List<HttpAuthenticationMechanism> httpAuthenticationMechanisms,
             @ConfigProperty(name = "quarkus.security.events.enabled") boolean securityEventsEnabled) {
         this.securityEventHelper = new SecurityEventHelper<>(authSuccessEvent, authFailureEvent, AUTHENTICATION_SUCCESS,
                 AUTHENTICATION_FAILURE, beanManager, securityEventsEnabled);
         this.identityProviderManager = identityProviderManager;
-        List<HttpAuthenticationMechanism> mechanisms = new ArrayList<>();
-        for (HttpAuthenticationMechanism mechanism : httpAuthenticationMechanism) {
-            boolean found = false;
-            for (Class<? extends AuthenticationRequest> mechType : mechanism.getCredentialTypes()) {
-                for (IdentityProvider<?> i : providers) {
-                    if (i.getRequestType().equals(mechType)) {
-                        found = true;
-                        break;
-                    }
-                }
-                if (found) {
+
```

Comment:
IMHO none of these changes are necessary, for example, removing the basic authentication mechanism silently is a problem as well.
I believe that it would be useful to fail the build if the mechanism has a credential type but no matching identity provider is found. Currently this mechanism is not loaded and then users are chasing why.

And indeed if the mechanism does not have specific credential types it should be added anyway as is the case now.

Reply:
IMHO all we have to do is to fail at https://github.com/quarkusio/quarkus/blob/main/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L58 if some mechanism which does not have any credential types have gone to calling it and the number of providers is 0.

This guard plus failing the build if the HTTP mechanism with the declared cred types finds no matching providers

Reply:
> IMHO none of these changes are necessary, 

it is necessary because a test is failing (for reasons explained below)

>  for example, removing the basic authentication mechanism silently is a problem as well.

that was already happening, it is not a new thing, it is what is there now

> I believe that it would be useful to fail the build if the mechanism has a credential type but no matching identity provider is found.

I don't think it is precise description because the basic authentication mechanism is added by default in case someone defines `IdentityProvider` which such credentials. Which we can't know when we add basic auth mechanisim because it's a runtime thing. Therefore we add it just in case and if there is no provider, we don't use it.

> Currently this mechanism is not loaded and then users are chasing why.

Not true because it is not used, therefore no authentication is happening. Please see `NoConfiguredRealmsTestCase.testSecureAccessFailure` that tries to authenticate request and fails because there is no identity provider.


I don't say we need to keep this behavior, but what you propose are breaking changes and I want you to be aware of it before we do anything.

> And indeed if the mechanism does not have specific credential types it should be added anyway as is the case now.

that is fine for all but basic auth mechanism

Reply:
> IMHO all we have to do is to fail at https://github.com/quarkusio/quarkus/blob/main/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L58 if some mechanism which does not have any credential types have gone to calling it and the number of providers is 0.

That is already happening, so what you propose is to just always add all mechanism, but that brings troubles I described above

> This guard plus failing the build if the HTTP mechanism with the declared cred types finds no matching providers

We can't fail a build because we don't know credential types at a build time. We can fail during runtime when the authenticator is initialized but that is hardening of current behavior. Only thing that will change in all your proposals is adding exception instead of silently removing the mechanism because we already didn't require mechanisms to provide credential types.

Here https://github.com/quarkusio/quarkus/issues/38508 you say that _
HttpAuthenticationMechanism is ignored when no IdentyProvider is registered_ which is not true for scenario when no credential types are provided. I am really sorry, but I try to understand where is a improvement in a goal of _not having to implement identity provider_. What will change to current state?

Reply:
Sounds reasonable @sberyozkin , thanks!

Reply:
Thanks @michalvavrik

Reply:
FYI `QuarkusIdentityProviderManagerImpl` portion described above is already how it works.

Reply:
@michalvavrik Sorry for re-opening, yes, QuarkusIdentityProviderManagerImpl has been taken care of, but I thought we agreed to remove the deprecation  and fail if no matching IdentityProvider is available for non-empty HttpAuthenticationMechanism.getCredentialTypes (say we repeat the same error in `HttpAuthenticator`), instead of ignoring the non-matching mechanism

Reply:
Yes, I resolved commit before pushing changes, sorry. Please check out current state and let's start from there because current code is how I understand what we agreed on.

Anyway, I think stuff around basic auth mechanism might come back to bite us. But it is hardening, I agree.